src/main.rs

```diff
@@ -663,77 +798,269 @@ fn get_setup(args: &[String]) -> Setup {
+                    );
+                }
+
+                let threshold = threshold.parse::<f64>().unwrap_or_else(|_| {
+                    error!("Invalid --{}: {}", NORMALISATION_THRESHOLD, threshold);
+                    println!("Valid --{} values: -10.0 - 0.0", NORMALISATION_THRESHOLD);
+                    println!(
+                        "Default: {}",
+                        ratio_to_db(PlayerConfig::default().normalisation_threshold)
+                    );
+                    exit(1);
+                });
+
+                if !(-10.0..=0.0).contains(&threshold) {
+                    error!("Invalid --{}: {}", NORMALISATION_THRESHOLD, threshold);
+                    println!("Valid --{} values: -10.0 - 0.0", NORMALISATION_THRESHOLD);
+                    println!(
+                        "Default: {}",
+                        ratio_to_db(PlayerConfig::default().normalisation_threshold)
+                    );
+                    exit(1);
+                }
+
+                db_to_ratio(threshold)
             })
             .unwrap_or(PlayerConfig::default().normalisation_threshold);
+
         let normalisation_attack = matches
             .opt_str(NORMALISATION_ATTACK)
             .map(|attack| {
-                Duration::from_millis(attack.parse::<u64>().expect("Invalid attack value"))
+                if !normalisation {
+                    warn!(
+                        "Without the --{} flag --{} has no effect.",
+                        ENABLE_VOLUME_NORMALISATION, NORMALISATION_ATTACK
+                    );
+                }
+
+                let attack = attack.parse::<u64>().unwrap_or_else(|_| {
+                    error!("Invalid --{}: {}", NORMALISATION_ATTACK, attack);
+                    println!("Valid --{} values: 1 - 500", NORMALISATION_ATTACK);
+                    println!(
+                        "Default: {}",
+                        PlayerConfig::default().normalisation_attack.as_millis()
+                    );
+                    exit(1);
+                });
+
+                if !(1..=500).contains(&attack) {
+                    error!("Invalid --{}: {}", NORMALISATION_ATTACK, attack);
+                    println!("Valid --{} values: 1 - 500", NORMALISATION_ATTACK);
+                    println!(
+                        "Default: {}",
+                        PlayerConfig::default().normalisation_attack.as_millis()
+                    );
+                    exit(1);
+                }
+
+                Duration::from_millis(attack)
             })
             .unwrap_or(PlayerConfig::default().normalisation_attack);
+
         let normalisation_release = matches
             .opt_str(NORMALISATION_RELEASE)
             .map(|release| {
-                Duration::from_millis(release.parse::<u64>().expect("Invalid release value"))
+                if !normalisation {
+                    warn!(
+                        "Without the --{} flag --{} has no effect.",
```

Comment:
This warning is best also placed with the Alsa mixer options.

Reply:
This doesn't have anything really to do with the alsa mixer does it?

Reply:
No, I was thinking along specifying `--alsa-xxx` options without the Alsa mixer enabled.

Reply:
This a normalization limiter setting it has nothing to do with the alsa mixer.

Reply:
I know I am taking the “without x flag y has no effect” as an example. For consistency you should also do that for all the Alsa flags, that’s my point.

Reply:
The problem really is that if the alsa backend is compiled in you can select a different backend (ofc if it's also compiled in) and select the alsa volume method and the alsa setting will have an effect even though you're  not using the alsa backend.

Reply:
How's that? Now alsa specific options aren't even evaluated unless the alsa backend is compiled in. The user just gets a warning of  "Alsa specific options have no effect if the alsa backend is not enabled at build time." if they try to use any.

Reply:
There. @roderickvd I think that about covers it as far as warnings. I paired the warnings down a little bit actually but no options that would logically do nothing, like normalization options if normalization is not enabled are not even evaluated but instead just set to their defaults.

Reply:
So as a recap:
 1. We're calling the ability to use alsa volume control with other backends a feature not a bug even though it may be unintentional and treating it as such and not throwing errors in that case.
 
 2. Options that would logically have no effects are not evaluated, default values are used, but the user is warned that such options have no effect if they try to use them.
 
 3. If an option is evaluated and an error happens the user is given a useful error message that tells them the valid values for the option and the default if applicable and librespot exits with an error.
 
 4. The options and flags have been cleaned up both order and formatting wise so main looks nicer and so the help menu is prettier.
 
 5. Some other misc clean up here and there.

Reply:
Ok @roderickvd I feature gated the error messages and option descriptions so for example here's what `--help` looks like with just the PulseAudio backend complied in:

```
Usage: ./librespot [options]

Options:
    -h, --help          Print this help menu.
    -v, --verbose       Enable verbose log output.
    -V, --version       Display librespot version string.
        --disable-audio-cache 
                        Disable caching of the audio data.
        --disable-credential-cache 
                        Disable caching of credentials.
        --disable-discovery 
                        Disable zeroconf discovery mode.
        --disable-gapless 
                        Disable gapless playback.
        --emit-sink-events 
                        Run PROGRAM set by `--onevent` before the sink is
                        opened and after it is closed.
        --enable-volume-normalisation 
                        Play all tracks at approximately the same apparent
                        volume.
        --autoplay      Automatically play similar songs when your music ends.
        --passthrough   Pass a raw stream to the output. Only works with the
                        pipe and subprocess backends.
    -n, --name NAME     Device name. Defaults to Librespot.
    -b, --bitrate BITRATE
                        Bitrate (kbps) {96|160|320}. Defaults to 160.
    -u, --username USERNAME
                        Username used to sign in with.
    -p, --password PASSWORD
                        Password used to sign in with.
    -c, --cache PATH    Path to a directory where files will be cached.
    -m, --mixer MIXER   Not supported by the current audio backend(s).
        --device-type TYPE
                        Displayed device type. Defaults to speaker.
        --system-cache PATH
                        Path to a directory where system files (credentials,
                        volume) will be cached. May be different from the
                        `--cache` option value.
        --cache-size-limit SIZE
                        Limits the size of the cache for audio files. It's
                        possible to use suffixes like K, M or G, e.g. 16G for
                        example.
        --onevent PROGRAM
                        Run PROGRAM when a playback event occurs.
        --proxy URL     HTTP proxy to use when connecting.
        --ap-port PORT  Connect to an AP with a specified port 1 - 65535. If
                        no AP with that port is present a fallback AP will be
                        used. Available ports are usually 80, 443 and 4070.
        --backend NAME  Audio backend to use. Use ? to list options.
        --device NAME   Not supported by the current audio backend(s).
        --format FORMAT Output format {F64|F32|S32|S24|S24_3|S16}. Defaults to
                        S16.
        --dither DITHER Specify the dither algorithm to use
                        {none|gpdf|tpdf|tpdf_hp}. Defaults to tpdf for formats
                        S16, S24, S24_3 and none for other formats.
        --mixer-name    
        --alsa-mixer-control NAME
                        Not supported by the current audio backend(s).
        --mixer-card    
        --alsa-mixer-device DEVICE
                        Not supported by the current audio backend(s).
        --mixer-index   
        --alsa-mixer-index NUMBER
                        Not supported by the current audio backend(s).
        --initial-volume VOLUME
                        Initial volume in % from 0 - 100. Defaults to 50.
        --zeroconf-port PORT
                        The port the internal server advertises over zeroconf
                        1 - 65535. Ports <= 1024 may require root privileges.
        --normalisation-method METHOD
                        Specify the normalisation method to use
                        {basic|dynamic}. Defaults to dynamic.
        --normalisation-gain-type TYPE
                        Specify the normalisation gain type to use
                        {track|album|auto}. Defaults to auto.
        --normalisation-pregain PREGAIN
                        Pregain (dB) applied by volume normalisation from
                        -10.0 to 10.0. Defaults to 0.0.
        --normalisation-threshold THRESHOLD
                        Threshold (dBFS) at which point the dynamic limiter
                        engages to prevent clipping from 0.0 to -10.0.
                        Defaults to -2.0.
        --normalisation-attack TIME
                        Attack time (ms) in which the dynamic limiter reduces
                        gain from 1 to 500. Defaults to 5.
        --normalisation-release TIME
                        Release or decay time (ms) in which the dynamic
                        limiter restores gain from 1 to 1000. Defaults to 100.
        --normalisation-knee KNEE
                        Knee steepness of the dynamic limiter from 0.0 to 2.0.
                        Defaults to 1.0.
        --volume-ctrl VOLUME_CTRL
                        Volume control scale type {cubic|fixed|linear|log}.
                        Defaults to log.
        --volume-range RANGE
                        Range of the volume control (dB) from 0.0 to 100.0.
                        Defaults to 60.0.
```

Reply:
Notice no mention of softvol vs alsa and the alsa specific options say `Not supported by the current audio backend(s).`

Reply:
Does look much nicer!